flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksDBSnapshotStrategyBase.java

```diff
@@ -325,6 +329,7 @@ protected RocksDBSnapshotOperation(
             this.checkpointStreamFactory = checkpointStreamFactory;
             this.stateMetaInfoSnapshots = stateMetaInfoSnapshots;
             this.localBackupDirectory = localBackupDirectory;
+            this.tmpResourcesRegistry = new CloseableRegistry();
```

Comment:
Can `tmpResourcesRegistry` be reused by multiple checkpoints?
Suppose chk-1 left some files because of the timeout, `tmpResourcesRegistry` would clean these files and close; if chk-2 also timeout again, `tmpResourcesRegistry.registerCloseable()` would throw `IOException` as `testUploadedSstCanBeCleanedUp()` shown, is this as expected?

Reply:
Hi @fredia , thanks a lot for your review. 

As I understand, the `tmpResourcesRegistry` cannot be reused by multiple checkpoints. If `tmpResourcesRegistry` is reused, it will be closed after chk-1 timeout, and chk-2 will call `tmpResourcesRegistry.registerCloseable()` after writing file. However, chk-2 will throw `IOException` even if chk-2 is fine.

Current PR doesn't reuse the `tmpResourcesRegistry` by multiple checkpoints, the `RocksDBSnapshotOperation` is created for single checkpoint, the constructor includes the checkpointId.

The `testUploadedSstCanBeCleanedUp` reuses `tmpResourcesRegistry` by multiple `uploader.uploadFilesToCheckpointFs`, it isn't reused by multiple checkpoints. A single checkpoint may call `uploader.uploadFilesToCheckpointFs` multiple times, for example: `RocksIncrementalSnapshotStrategy#uploadSstFiles`.

Please correct me if I'm wrong, thanks~ 🤔

<img width="995" alt="image" src="https://user-images.githubusercontent.com/38427477/212459586-7c7327a4-1b8c-4dcb-a16a-aa5bb56b9443.png">

Reply:
Thanks for the clarification, I forgot that the RocksDBSnapshotOperation is created for single checkpoint😊. 
And I have another question: will “`tmpResourcesRegistry.registerCloseable()` throw `IOException` after `close()`” cause the job to failover? If we set `execution.checkpointing.tolerable-failed-checkpoints` to infinity, if chk-1 timeout,  will the job continue to run?

Reply:
Thanks for your quick feedback!

> will “tmpResourcesRegistry.registerCloseable() throw IOException after close()” cause the job to failover?

It doesn't cause job to failover directly, it just causes current checkpoint fails. 

We can see the code, `outputStream.closeAndGetHandle();` and `tmpResourcesRegistry.registerCloseable()` are in the same method, they are executed in uploader thread. So the `tmpResourcesRegistry.registerCloseable()` throws exception is similar to `outputStream.closeAndGetHandle();`.

As we know, if the [execution.checkpointing.tolerable-failed-checkpoints](https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/deployment/config/#execution-checkpointing-tolerable-failed-checkpoints) > 0, the job will continue to run after writing checkpoint file fails. So `tmpResourcesRegistry.registerCloseable()` is also work.

<img width="998" alt="image" src="https://user-images.githubusercontent.com/38427477/212467318-27036f4f-9001-4cfa-abdc-e741f27dd8a6.png">